nifi-extension-bundles/nifi-amqp-bundle/nifi-amqp-processors/src/main/java/org/apache/nifi/amqp/processors/PublishAMQP.java

```diff
@@ -158,10 +188,21 @@ protected void processResource(final Connection connection, final AMQPPublisher
             throw new IllegalArgumentException("Failed to determine 'routing key' with provided value '"
                 + context.getProperty(ROUTING_KEY) + "' after evaluating it as expression against incoming FlowFile.");
         }
+        String selectedHeaderSource = context.getProperty(HEADERS_SOURCE).getValue();
```

Comment:
This should use `asDescribedValue()` and return `HeaderSource` instead of the string value.